File: src/main/java/pl/patrykdolata/lifemanager/security/JwtTokenProvider.java
package pl.patrykdolata.lifemanager.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.User;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.security.Key;
import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import java.util.stream.Collectors;

@Component
public class JwtTokenProvider {

    private final Logger log = LoggerFactory.getLogger(JwtTokenProvider.class);
    private static final String AUTHORITIES_KEY = "auth";
    private static final String ID_KEY = "id";

    private Key key;

    @Value("${jwt.expiration-time}")
    private long tokenExpirationTime;

    @Value("${jwt.remember-me-expiration-time}")
    private long rememberMeExpirationTime;

    @Value("${jwt.secret}")
    private String jwtSecret;

    @PostConstruct
    public void init() {
        byte[] keyBytes = Decoders.BASE64.decode(jwtSecret);
        this.key = Keys.hmacShaKeyFor(keyBytes);
    }

    public String createToken(Authentication authentication, boolean rememberMe) {
        long now = new Date().getTime();
        Date validity = rememberMe ? new Date(now + rememberMeExpirationTime) : new Date(now + tokenExpirationTime);

        return Jwts.builder()
                .setSubject(authentication.getName())
                .claim(AUTHORITIES_KEY, getAuthorities(authentication))
                .claim(ID_KEY, getId(authentication))
                .signWith(key, SignatureAlgorithm.HS512)
                .setExpiration(validity)
                .compact();
    }

    public Authentication getAuthentication(String token) {
        Claims claims = getTokenClaims(token);
        Collection<GrantedAuthority> authorities = getAuthorities(claims);
        Long id = getId(claims);
        AuthenticatedUser principal = new AuthenticatedUser(id, claims.getSubject(), "", true, authorities);

        return new UsernamePasswordAuthenticationToken(principal, token, authorities);
    }

    private Claims getTokenClaims(String token) {
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    private Long getId(Authentication authentication) {
        return ((AuthenticatedUser) authentication.getPrincipal()).getId();
    }

    private String getAuthorities(Authentication authentication) {
//        return authentication.getAuthorities().stream()
//                .map(GrantedAuthority::getAuthority)
//                .collect(Collectors.joining(","));
        return "ROLE_USER";
    }

    private Collection<GrantedAuthority> getAuthorities(Claims claims) {
        return Arrays.stream(claims.get(AUTHORITIES_KEY).toString().split(","))
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toList());
    }

    private Long getId(Claims claims) {
        return Long.valueOf(claims.get(ID_KEY).toString());
    }

    public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
            return true;
        } catch (JwtException | IllegalArgumentException e) {
            log.debug("Invalid JWT token");
        }
        return false;
    }
}

File: src/main/kotlin/pl/patrykdolata/lifemanager/security/JwtTokenProvider.kt
package pl.patrykdolata.lifemanager.security

import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Component
import java.security.Key
import java.util.*
import javax.annotation.PostConstruct

@Component
class JwtTokenProvider(
        @Value("\${jwt.expiration-time}")
        private val tokenExpirationTime: Long,

        @Value("\${jwt.remember-me-expiration-time}")
        private val rememberMeExpirationTime: Long,

        @Value("\${jwt.secret}")
        private val jwtSecret: String
) {

    companion object {
        private const val AUTHORITIES_KEY = "auth"
        private const val ID_KEY = "id"
    }

    private val log: Logger = LoggerFactory.getLogger(JwtTokenProvider::class.java)

    private lateinit var key: Key

    @PostConstruct
    fun init() {
        val keyBytes: ByteArray = Decoders.BASE64.decode(jwtSecret)
        key = Keys.hmacShaKeyFor(keyBytes)
    }

    fun createToken(authentication: Authentication, rememberMe: Boolean = false): String {
        val now: Long = Date().time
        val validity: Date = if (rememberMe) Date(now + rememberMeExpirationTime) else
            Date(now + tokenExpirationTime)
        return Jwts.builder()
                .setSubject(authentication.name)
                .claim(AUTHORITIES_KEY, getAuthorities(authentication))
                .claim(ID_KEY, getId(authentication))
                .signWith(key, SignatureAlgorithm.HS512)
                .setExpiration(validity)
                .compact()
    }

    fun getAuthentication(token: String): Authentication {
        val claims: Claims = getTokenClaims(token)
        val authorities = getAuthorities(claims)
        val id: Long = getId(claims)
        val principal = AuthenticatedUser(id, claims.subject)

        return UsernamePasswordAuthenticationToken(principal, token, authorities)
    }

    fun validateToken(token: String): Boolean {
        try {
            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token)
            return true
        } catch (e: Exception) {
            when (e) {
                is JwtException, is IllegalArgumentException -> log.debug("Invalid JWT token")
            }
        }
        return false
    }

    private fun getTokenClaims(token: String): Claims {
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .body
    }

    private fun getId(authentication: Authentication): Long = (authentication.principal as AuthenticatedUser).id

    private fun getAuthorities(authentication: Authentication) = authentication.authorities.joinToString(";")

    private fun getAuthorities(claims: Claims): Collection<GrantedAuthority> =
            claims[AUTHORITIES_KEY].toString().split(",").map { SimpleGrantedAuthority(it) }


    private fun getId(claims: Claims): Long = claims[ID_KEY].toString().toLong()
}
